### DISCUSSION TITLE
toolkit: ARM64

### DISCUSSION THREAD

**Discussion by moikot** [fluxcd/flux2#188] [Proposals] [answered]
I'm looking for a solution for reconciling the state of my cluster with Git repo. I want, literally, to provision my cluster, install some essential components, like parts of FluxCD toolkit and some secret management tool Hashicorp vault, unseal the vault, point to source-controller to the sources, and everything else should be installed automatically based on the configuration I keep in Git. 

I know that I can install source-controller, kustomize-controller, and helm-controller (since all of my services are using Helm) and create CRDs describing the resources I want to monitor and reconcile. The problem is that CRDs like Kustomization and HelmRelease have to be present in the cluster before the reconciliation process could start. I can potentially create one Kustomization CRD pointing to a GIT repo with all other HelmReleases needed for the rest of my services and I hope kustomize-controller will create CRDs and helm-controller reconcile them. 

I have to mention here that my cluster is running on arm64 SBCs and the problem with kustomize-controller is that it relies on kustomize and kubectl pulled as binaries and although kubectl can be pulled for arm64, kustomize fails to provide such option. I can easily produce multi-arch images for source-controller and helm-controller but not for kustomize-contoller.

I've just started thinking if I can implement a very stupid controller which just takes all YAML files from a folder in a git repo and applies them to a Kubernetes cluster. Or maybe I'm missing some other option here?

**Comment by stefanprodan** [comment, marked as answer]
> I can potentially create one Kustomization CRD pointing to a GIT repo with all other HelmReleases needed for the rest of my services and I hope kustomize-controller will create CRDs and helm-controller reconcile them.

The `tk bootstrap` command does the initial setup for you.

> I have to mention here that my cluster is running on arm64 SBCs and the problem with kustomize-controller is that it relies on kustomize

We need to drop the kustomize binary and rewrite the controller parts that are shelling out to it. Fixed in https://github.com/fluxcd/kustomize-controller/pull/92

**Reply by moikot** [reply]
@stefanprodan that was a really quick turnaround! Thanks a lot!
I feel like I'm asking too much, but do you think it's possible to get rid of kubectl dependency? The server-side apply might suffice, I'm not an expert though.

What I can do is to help with multi-arch builds, if you need any help at all. In my fork of source-controller, I modified [release.yml](https://github.com/moikot/source-controller/blob/master/.github/workflows/release.yml), where I'm using [build push action v2](https://github.com/docker/build-push-action/pull/92) which is going to be released very soon I believe. You can find the results for the build [here](https://hub.docker.com/r/moikot/source-controller/tags). As soon as build push action v2 released I can prepare PRs.

**Reply by stefanprodan** [reply]
Removing kubectl is not an option, there must be some command that we can use the Dockerfile to get the ARCH and use it to compose the kubectl download URL.

**Reply by moikot** [reply]
I guess it's OK to fetch kubectl binaries using TARGETPLATFORM argument since platform/architecture names should match.

**Reply by stefanprodan** [reply]
Yes that works, I've started working on ARM64 builds for all the toolkit controllers: https://github.com/fluxcd/toolkit/issues/194

**Reply by moikot** [reply]
That's great! Are you sure you don't want to have the same name for the images using multi-platform manifests? By using them it is possible to avoid --arch=arm64 flag and they are supported out of the box by buildx.

**Reply by stefanprodan** [reply]
GitHub doesn't have support for multi-arch images and we're moving away from Docker Hub.

**Reply by moikot** [reply]
I agree that Docker Hub is becoming less and less attractive. And I might be wrong, but it seems like manifest list v2 going to be rolled out [quite soon](https://github.com/containerd/containerd/issues/4520#issue-690523432) on GHCR, probably at the same time when the build push action v2 is released.

**Reply by stefanprodan** [reply]
@moikot can you please test the latest gotk release on ARM64?

```
gotk install --arch=arm64 
```

**Reply by moikot** [reply]
@stefanprodan I can confirm that all four controllers are working just fine in my arm64 cluster. Thanks a lot for making it possible!

**Reply by stefanprodan** [reply]
Thank you for testing it, much appreciate 👍
